### DISCUSSION TITLE
Change `{% dd %}` and `{{ dump }}` to be syntactically and functionally consistent

### DISCUSSION THREAD

**Discussion by aaronbushnell** [craftcms/cms#12479] [Ideas] [answered]
Related to #10612—

Right now there's a handful of different "dump" methods with varying functions:
- `Craft::dd()`: Dump and die; uses Symfony VarDumper and then exits
- `Craft::dump()`: Dump; uses Symfony VarDumper
- `{% dd %}`: Dump and die; calls `Craft::dd()`
- `{{ dump() }}`: Baked into Twig; calls PHP's `var_dump()`

While the PHP calls make sense to me, I think the Twig versions should be made consistent in syntax and output style:
- Override the `{{ dump() }}` function to call `Craft::dump()`
- Change `{% dd %}` to `{{ dd }}` (which calls `Craft::dd()`)
- Optionally keep `{% dd %}` for a bit and flag it as a deprecated Twig token to be removed eventually

Here are the changes in use:
```twig
{# Twig functions that both use Symfony VarDumper via Craft::dump() and Craft::dd() respectively #}
{{ dump(entry) }}
{{ dd(entry) }}

{# Deprecated, removed in a future Craft version #}
{% dd(entry) %}
```

**Comment by brandonkelly** [comment, marked as answer]
Agree that `dump()` should go through `Craft::dump()`. And probably should not be reliant on Dev Mode.

It doesn’t make sense to change `{% dd %}` to `{{ dd() }}` though, because it doesn’t output anything inline in the template like `dump()` does.

**Reply by aaronbushnell** [reply]
That’s fair! I was borrowing the syntax from Blade but it looks like it’s because the `dd` function there doesn’t halt the page from rendering entirely. It renders up to that point and then stops.

**Reply by brandonkelly** [reply]
Made the `dump()` changes for Craft 4.4 (#12486).

**Reply by aaronbushnell** [reply]
Thanks, @brandonkelly! Enjoy the holidays!

**Reply by aaronbushnell** [reply]
I wonder if it'd also make sense to change `Craft::dd()` to use Symfony's VarDumper, too? Here's a before/after. I think this makes it a bit easier to peruse the values in the output. Unless Yii's baked-in `dump` function has a more useful output of Yii things?

![Screenshot 2022-12-27 at 8 52 08 AM](https://user-images.githubusercontent.com/315202/209676451-2b10b933-42a7-4134-953f-18f341aad7af.png)

**Reply by brandonkelly** [reply]
Good idea. Just made that change for 4.4 as well (daae067a28e75671ae94a42cd0773fd0ebac08e0).

**Reply by brandonkelly** [reply]
4.4 is out with those changes!

**Reply by aaronbushnell** [reply]
This is great! Thanks, Brandon!

**Reply by khalwat** [reply]
All you need to do is Control-click on the arrow to expand all, if anyone wants that
